operator/cmd/mesh/install.go

```diff
@@ -348,3 +369,82 @@ func getProfileNSAndEnabledComponents(iop *v1alpha12.IstioOperator) (string, str
 	}
 	return iop.Spec.Profile, name.IstioDefaultNamespace, enabledComponents, nil
 }
+
+// requiresIstiodServiceCreation determines if the CP requires an istiod service to function
+func requiresIstiodServiceCreation(cs kubernetes.Interface, istioNs string) bool {
+	// First: does the istiod service exist?
+	_, err := cs.CoreV1().Services(istioNs).Get(
+		context.TODO(), "istiod", metav1.GetOptions{})
+	if kerrors.IsNotFound(err) {
+		vwh, err := cs.AdmissionregistrationV1().ValidatingWebhookConfigurations().Get(
+			context.TODO(), fmt.Sprintf("istiod-%s", name.IstioDefaultNamespace), metav1.GetOptions{})
+		if err != nil {
+			return false
+		}
+		// Second: if it doesn't, are there webhooks in the VWHC referencing it? If so,
+		// we should create to avoid broken validation
+		for _, wh := range vwh.Webhooks {
+			if wh.ClientConfig.Service != nil {
+				if wh.ClientConfig.Service.Name == "istiod" {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// createIstiodService creates an `istiod` service that selects ALL istiod instances.
+func createIstiodService(cs kubernetes.Interface, istioNs string) error {
+	const (
+		pilotDiscoveryChart = "istio-control/istio-discovery"
+		serviceTemplateName = "service.yaml"
+	)
+	r := helm.NewHelmRenderer("", pilotDiscoveryChart, "Pilot", istioNs)
+
+	if err := r.Run(); err != nil {
+		return fmt.Errorf("failed running Helm renderer: %v", err)
+	}
+
+	values := `
+revision: ""
+`
+
+	serviceYAML, err := r.RenderManifestFiltered(values, func(tmplName string) bool {
+		return strings.Contains(tmplName, serviceTemplateName)
+	})
+	if err != nil {
+		return fmt.Errorf("failed rendering service manifest: %v", err)
+	}
+
+	scheme := runtime.NewScheme()
+	codecFactory := serializer.NewCodecFactory(scheme)
+	deserializer := codecFactory.UniversalDeserializer
+
+	svcObject, _, err := deserializer().Decode([]byte(serviceYAML), nil, &corev1.Service{})
+	if err != nil {
+		return fmt.Errorf("failed deserializing generated manifest")
+	}
+
+	svc, ok := svcObject.(*corev1.Service)
+	if !ok {
+		return fmt.Errorf("failed casting deserialized manifest as service")
+	}
+
+	// Rename the generated service from `istiod-<revision>` to `istiod`
+	svc.Name = "istiod"
+
+	// Do not remove this service when the revision is removed
+	delete(svc.Labels, label.IoIstioRev.Name)
+
+	// Change selectors such that the `istiod` service selects istiod instances from all revisions
+	svc.Spec.Selector = map[string]string{
```

Comment:
Is this needed?  I assume the `app: Istiod` label remains after you delete the istio rev label?

Reply:
Yeah so without this, the selectors will be `app: istiod, istio: pilot` which won't select revisions, just default installs. To get it selecting all revisions we just want `app: istiod`